docs/guides/working-with-json/json-semi-structured.md

```diff
@@ -0,0 +1,459 @@
+   FROM github_json
+   WHERE (event.type = 'WatchEvent') AND (toYear(parseDateTimeBestEffort(event.created_at)) >= 2022) AND (repo IN
+   (
+       SELECT lower(event.repo.name) AS repo
+       FROM github_json
+       WHERE (event.type = 'WatchEvent') AND (toYear(parseDateTimeBestEffort(event.created_at)) >= 2022)
+       GROUP BY event.repo.name
+       ORDER BY count() DESC
+       LIMIT 10
+   ))
+   GROUP BY
+       repo,
+       hour
+)
+GROUP BY repo
+ORDER BY repo ASC;
+```
+
+## Adding Primary Keys
+
+The above example is not realistic in that it has no primary or sort key i.e., it uses `tuple()`. This negates the benefit of the index features in ClickHouse. To add a primary key, and still exploit the JSON object capabilities, we recommended using a dedicated subkey for the JSON. This requires inserting the data using the JSONAsRow format instead of JSONAsObject. For example, consider the JSON below and the corresponding table definition and insert statement.
+
+```sql
+SET allow_experimental_object_type=1;
+
+DROP TABLE IF EXISTS github_json;
+
+CREATE table github_json
+(
+   event_type Enum('CommitCommentEvent' = 1, 'CreateEvent' = 2, 'DeleteEvent' = 3,
+   'ForkEvent' = 4, 'GollumEvent' = 5, 'IssueCommentEvent' = 6, 'IssuesEvent' = 7, 'MemberEvent' = 8, 
+   'PublicEvent' = 9, 'PullRequestEvent' = 10, 'PullRequestReviewCommentEvent' = 11, 
+   'PushEvent' = 12, 'ReleaseEvent' = 13, 'SponsorshipEvent' = 14, 'WatchEvent' = 15, 'GistEvent' = 16, 'FollowEvent' = 17, 'DownloadEvent' = 18, 'PullRequestReviewEvent' = 19,
+   'ForkApplyEvent' = 20, 'Event' = 21, 'TeamAddEvent' = 22),
+    repo_name LowCardinality(String),
+    event      JSON
+) ENGINE = MergeTree ORDER BY (event_type, repo_name);
+```
+
+Inserting data requires us to use the JSONEachRow format. Note how the `event` sub field now holds our dynamic JSON, whilst the root keys are explicitly defined.
+
+```sql
+INSERT INTO github_json FORMAT JSONEachRow
+{"event":{"type":"PushEvent","actor":{"avatar_url":"https://avatars.githubusercontent.com/u/41898282?",
+"display_login":"github-actions","gravatar_id":"","id":41898282,"login":"github-actions[bot]",
+"url":"https://api.github.com/users/github-actions[bot]"},"repo":{"id":410071248,
+"name":"pioug/yield-data","url":"https://api.github.com/repos/pioug/yield-data"}},
+"event_type":"PushEvent","repo_name":"pioug/yield-data"}
+```
+
+This requires a restructuring of our JSON, which is inconvenient at best. Ideally, we need a more flexible approach that allows us to modify the fields we wish to extract as root keys over time without needing to change our data pipelines. Inserting our row as a `String` inside an EPHEMERAL column `message_raw`, we can extract specific fields of interest using DEFAULT expressions for the root fields. The `String` EPHEMERAL column is also mapped to a JSON object column `message` that provides the usual flexibility. This EPHEMERAL column will not be persisted and will be discarded at INSERT time.
```

Comment:
We can also mention here that fields we had to add explicitly for the PK hold redundant data already present in the JSON

Reply:
We can use a link to the docs about EPHEMERAL:
https://clickhouse.com/docs/en/sql-reference/statements/create/table/#ephemeral

Reply:
added Our primary key fields are as a result duplicated i.e. they occur at the root of the document, as well as in the `message` JSON.

Reply:
added link thanks